### ISSUE TITLE
tr/// can write on immortals

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14482]
Migrated from [rt.perl.org#123759](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123759) (status was 'resolved')

Searchable as RT123759$

**Comment by p5pRT** [comment]
### From @hvds

#### Created by @hvds



Here's another finding from AFL&#8203;:

% perl \-e '0=~m\{\}=~y///d'
Segmentation fault \(core dumped\)
% gdb perl core
\[\.\.\.\]
Program received signal SIGSEGV\, Segmentation fault\.
0x0000000000692682 in S\_do\_trans\_complex \(sv=0xa5bc90 \<PL\_sv\_yes>\)
&nbsp;    at doop\.c&#8203;:178
178                         \*d\+\+ = \*s;
\(gdb\) p d
$1 = \(U8 \*\) 0x82eb24 \<PL\_Yes\+1> ""
\(gdb\) p s
$2 = \(U8 \*\) 0x82eb23 \<PL\_Yes> "1"
\(gdb\) where
\#0  0x0000000000692682 in S\_do\_trans\_complex \(sv=0xa5bc90 \<PL\_sv\_yes>\)
&nbsp;    at doop\.c&#8203;:178
\#1  0x0000000000695351 in Perl\_do\_trans \(sv=0xa5bc90 \<PL\_sv\_yes>\) at doop\.c&#8203;:661
\#2  0x0000000000608372 in Perl\_pp\_trans \(\) at pp\.c&#8203;:765
\#3  0x000000000054514b in Perl\_runops\_debug \(\) at dump\.c&#8203;:2231
\#4  0x000000000046048b in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:2423
\#5  0x000000000045facf in perl\_run \(my\_perl=0xa5d010\) at perl\.c&#8203;:2346
\#6  0x000000000041eeb5 in main \(argc=2\, argv=0x7fffffffe648\, 
&nbsp;    env=0x7fffffffe660\) at perlmain\.c&#8203;:116

Bisect blames this&#8203;:

&nbsp;  commit 0d96b528ff0140192b273a677f66f80ee54f3fc4
&nbsp;  Author&#8203;: Chip Salzenberg \<chip@&#8203;pobox\.com>
&nbsp;  Date&#8203;:   Mon Jul 23 22&#8203;:40&#8203;:46 2012 \-0700

&nbsp;      Unify code that initializes constants yes\, no\, and undef

\.\. but the real problem is that the modified READONLY check in Perl\_do\_trans
assumes OPpTRANS\_IDENTICAL will always take us to S\_do\_trans\_count\{\,\_utf8\} \-
in this case we also have OPpTRANS\_DELETE\, so the end up in do\_trans\_complex
via the default case\.

I think the fix is to ensure we do end up at do\_trans\_count\_\*\(\)\, either
by canonicalizing the other flags when we set OPpTRANS\_IDENTICAL or by
fixing the overly tricksy Perl\_do\_trans case statement to something like&#8203;:

&nbsp;    if \(flags & OPpTRANS\_IDENTICAL\) \{
&nbsp;        return hasutf
&nbsp;            ? do\_trans\_count\_utf8\(sv\)
&nbsp;            : do\_trans\_count\(sv\);
&nbsp;    \} else if \(flags & \(OPpTRANS\_SQUASH|OPpTRANS\_DELETE|OPpTRANS\_COMPLEMENT\)\) \{
&nbsp;        return hasutf
&nbsp;            ? do\_trans\_complex\_utf8\(sv\)
&nbsp;            : do\_trans\_complex\(sv\);
&nbsp;    \} else \{
&nbsp;        return hasutf
&nbsp;            ? do\_trans\_simple\_utf8\(sv\)
&nbsp;            : do\_trans\_simple\(sv\);
&nbsp;    \}

\(or maybe both\, for extra clarity\)\.

Incidentally\, AFL originally found this using nul bytes as the y///
delimiters\, are we intentionally supporting that?

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.21.9:

Configured by hv at Sat Feb  7 18:19:55 GMT 2015.

Summary of my perl5 (revision 5 version 21 subversion 9) configuration:
  Commit id: 3c47da3c2ebf51f08bb927dfa456939cc6c8c30f
  Platform:
    osname=linux, osvers=3.13.0-37-generic, archname=x86_64-linux
    uname='linux shad2 3.13.0-37-generic #64-ubuntu smp mon sep 22 21:28:38 utc 2014 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Dcc=gcc -Dprefix=/opt/blead-d -Doptimize=-g -O6 -DDEBUGGING -Dusedevel -Uversiononly'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g -O6',
    cppflags='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.8.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/4.8/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.19.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.19'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -O6 -L/usr/local/lib -fstack-protector'



@INC for perl 5.21.9:
    /opt/blead-d/lib/perl5/site_perl/5.21.9/x86_64-linux
    /opt/blead-d/lib/perl5/site_perl/5.21.9
    /opt/blead-d/lib/perl5/5.21.9/x86_64-linux
    /opt/blead-d/lib/perl5/5.21.9
    .


Environment for perl 5.21.9:
    HOME=/home/hv
    LANG=C
    LANGUAGE=en_GB:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/hv/bin:/home/hv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Feb 08 06&#8203;:22&#8203;:02 2015\, hv wrote&#8203;:
> Incidentally\, AFL originally found this using nul bytes as the y///
> delimiters\, are we intentionally supporting that?

We \*depend\* on that\.  Search for ‘NUL byte as a the delimiter’ \(typos and all\) in perl\.c\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @hvds

Now fixed by a53bfdae91&#8203;:

&nbsp;    \[perl \#123759\] always count on OPpTRANS\_IDENTICAL
&nbsp;    
&nbsp;    If we detect that an in\-place transliteration will not result in any
&nbsp;    changes to the string\, we set OPpTRANS\_IDENTICAL and skip the normal
&nbsp;    checks for readonlyness; but if we do that\, we must make sure to use
&nbsp;    the same logic to decide which transliteration strategy to use\, or
&nbsp;    we may end up trying to write to the readonly string anyway\.
&nbsp;    
&nbsp;    This resulted in several ways to hit assert failures\, found by AFL
&nbsp;    \(\<http&#8203;://lcamtuf\.coredump\.cx/afl>\)\.

Hugo

**Comment by p5pRT** [comment]
@hvds - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
